### ISSUE TITLE
core dump when running `iec-identify.nse` script

### ISSUE THREAD

**Issue by omergunal** [nmap/nmap#1302]
System: 16.04
```
nmap --version
Nmap version 7.70SVN ( https://nmap.org )
Platform: x86_64-unknown-linux-gnu
Compiled with: nmap-liblua-5.3.4 openssl-1.1.0h nmap-libssh2-1.8.0 libz-1.2.8 nmap-libpcre-7.6 nmap-libpcap-1.8.1 nmap-libdnet-1.12 ipv6
Compiled without:
Available nsock engines: epoll poll select
```

```
sudo nmap -sV --script iec-identify.nse 192.168.1.109 -d
Starting Nmap 7.70SVN ( https://nmap.org ) at 2018-08-09 21:09 +03
PORTS: Using top 1000 ports found open (TCP:1000, UDP:0, SCTP:0)
--------------- Timing report ---------------
  hostgroups: min 1, max 100000
  rtt-timeouts: init 1000, min 100, max 10000
  max-scan-delay: TCP 1000, UDP 1000, SCTP 1000
  parallelism: min 0, max 0
  max-retries: 10, host-timeout: 0
  min-rate: 0, max-rate: 0
---------------------------------------------
NSE: Using Lua 5.3.
NSE: Arguments from CLI: 
NSE: Loaded 45 scripts for scanning.
NSE: Script Pre-scanning.
NSE: Starting runlevel 1 (of 2) scan.
Initiating NSE at 21:09
Completed NSE at 21:09, 0.00s elapsed
NSE: Starting runlevel 2 (of 2) scan.
Initiating NSE at 21:09
Completed NSE at 21:09, 0.00s elapsed
mass_rdns: Using DNS server 127.0.1.1
Initiating Parallel DNS resolution of 1 host. at 21:09
mass_rdns: 0.03s 0/1 [#: 1, OK: 0, NX: 0, DR: 0, SF: 0, TR: 1]
Completed Parallel DNS resolution of 1 host. at 21:09, 0.03s elapsed
DNS resolution of 1 IPs took 0.03s. Mode: Async [#: 1, OK: 0, NX: 1, DR: 0, SF: 0, TR: 1, CN: 0]
Initiating SYN Stealth Scan at 21:09
Scanning 192.168.1.109 [1000 ports]
Packet capture filter (device lo): dst host 192.168.1.109 and (icmp or icmp6 or ((tcp or udp or sctp) and (src host 192.168.1.109)))
Discovered open port 80/tcp on 192.168.1.109
Discovered open port 443/tcp on 192.168.1.109
Discovered open port 9091/tcp on 192.168.1.109
Discovered open port 902/tcp on 192.168.1.109
Increased max_successful_tryno for 192.168.1.109 to 1 (packet drop)
Completed SYN Stealth Scan at 21:09, 1.61s elapsed (1000 total ports)
Overall sending rates: 658.62 packets / s, 28979.10 bytes / s.
Segmentation fault (core dumped)
```

**Comment by dmiller-nmap** [comment]
It's possible this is a duplicate of #1147, where certain service match lines were resulting in stack exhaustion. To be sure, can you run the following command and show the output?

    nmap -sV 192.168.1.109 -d2 --version-trace

Alternatively, since I don't see the "Initiating Service scan" output line, it could be that there is some other crash. If you can generate a core file, it would greatly help us debug the problem:

1. Save your old core dump settings: `cat /proc/sys/kernel/core_pattern > /tmp/old-core-pattern`
2. Enable core dump to file: `echo "core.%t" | sudo tee /proc/sys/kernel/core_pattern`
3. Generate a crash dump: `nmap -sV 192.168.1.109 -d2 --version-trace`
4. Collect the `core.<timestamp>` file and send it to dev@nmap.org or dmiller@nmap.org
5. Restore the old core dump setting: `cat /tmp/old-core-pattern | sudo tee /proc/sys/kernel/core_pattern`

**Comment by dmiller-nmap** [comment]
To clarify, it does not seem that the problem is in the `iec-identify` NSE script, since that script would not be run and the debug output shows the crash happens before the version detection scan starts.

**Comment by sfx2000** [comment]
For OP - check to ensure **libpcre3-dev** is present (for debian/ubuntu) - was getting segfaults building nmap off the svn tip until I checked and only **libpcre3** was installed...

**Comment by omergunal** [comment]
@sfx2000  should i remove liprce3?
@dmiller-nmap i sent mail to you for logs

**Comment by sfx2000** [comment]
@omergunal - just add libpcre3-dev and keep libpcre3, and you should be fine...

**Comment by dmiller-nmap** [comment]
Ok, I've reproduced the problem by configuring `--with-libpcre=included`. I'll update when I figure out what is going wrong. Probably related to the fix for #1147

**Comment by dmiller-nmap** [comment]
According to PCRE API documentation:

> If studying the pattern does not produce any useful information, pcre_study() returns NULL by default. In that circumstance, if the calling program wants to pass any of the other fields to pcre_exec() or pcre_dfa_exec(), it must set up its own pcre_extra block. However, if pcre_study() is called with the PCRE_STUDY_EXTRA_NEEDED option, it returns a pcre_extra block even if studying did not find any additional information. It may still return NULL, however, if an error occurs in pcre_study(). 

I was working on the assumption that `pcre_study` always returned a `pcre_extra`, but this wasn't the case. I'm really not sure why it wasn't crashing every time, but apparently some versions of PCRE *do* return a pcre_extra every time. In any case, I'm adding requisite code to ensure that `regex_extra` is not NULL, since we are using it every time now.
